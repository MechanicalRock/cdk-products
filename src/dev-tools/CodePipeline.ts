import { Pipeline, Stage } from '@aws-cdk/aws-codepipeline';
import { PipelineBuildAction, PipelineProject, CodeCommitSource, ComputeType, LinuxBuildImage } from '@aws-cdk/aws-codebuild';
import { PipelineSourceAction, Repository } from '@aws-cdk/aws-codecommit';
import { Stack } from '@aws-cdk/cdk';
import { Role, ServicePrincipal } from '@aws-cdk/aws-iam';

export interface CodePipelineProps {
    stack: Stack,
    logicalName: string,
    pipelineName: string,
    codeRepository: Repository
}

export class CodePipeline {
    pipeline: Pipeline;
    constructor(private props: CodePipelineProps) {
        this.create();
    }

    create = () => {
        // Create the pipeline
        const pipeline = new Pipeline(this.props.stack, this.props.logicalName, {
            pipelineName: this.props.pipelineName,
            restartExecutionOnUpdate: true
        });
        this.pipeline = pipeline;

        const codeBuildRole = this.createCodeBuildRole();
        this.createSourceStage();
        this.createDevStage(codeBuildRole);
        this.createProdStage(codeBuildRole);
    }

    private createCodeBuildRole = (): Role => {
        const codeBuildProjectRole = new Role(this.props.stack, 'CodeBuildRole', {
            assumedBy: new ServicePrincipal('codebuild.amazonaws.com')
        });
        codeBuildProjectRole.attachManagedPolicy('arn:aws:iam::aws:policy/AdministratorAccess');
        return codeBuildProjectRole;
    }

    private createProdStage = (role: Role) => {
        // Prod stage and build & test action
        const prodStage = new Stage(this.props.stack, 'ProdStage', {
            pipeline: this.pipeline
        });

        const prodCodeBuildProject = this.createCodeBuildProject(role, 'DeployToProd');

        new PipelineBuildAction(this.props.stack, 'prodDeployAction', {
            project: prodCodeBuildProject,
            stage: prodStage
        })
    }

    private createDevStage = (role: Role) => {
        // Dev stage and build & test action
        const devStage = new Stage(this.props.stack, 'DevStage', {
            pipeline: this.pipeline
        });

        const buildAndTestCodeBuildProject = this.createCodeBuildProject(role, 'BuildAndTest');

        new PipelineBuildAction(this.props.stack, 'BuildAndTestAction', {
            project: buildAndTestCodeBuildProject,
            stage: devStage
        })
    }

    private createCodeBuildProject = (role: Role, logicalName: string): PipelineProject => {
        return new PipelineProject(this.props.stack, logicalName, {
            environment: {
                buildImage: LinuxBuildImage.UBUNTU_14_04_NODEJS_8_11_0
            },
            role
        })
    }

    private createSourceStage = () => {
        // Source stage and action
        const sourceStage = new Stage(this.props.stack, 'SourceStage', {
            pipeline: this.pipeline,
        });

        new PipelineSourceAction(this.props.stack, 'SourceAction', {
            repository: this.props.codeRepository,
            stage: sourceStage,
            pollForSourceChanges: true
        })
    }
}