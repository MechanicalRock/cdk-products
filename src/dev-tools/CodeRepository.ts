import { Repository } from '@aws-cdk/aws-codecommit';
import { Stack, Output } from '@aws-cdk/cdk';

export class CodeRepository {
    repository: Repository;

    constructor(private stack: Stack, private repoLogicalName: string, private repoName: string) {
        this.repository = this.create();
        this.createOutputs();
    }

    create(): Repository {
        return new Repository(this.stack, this.repoLogicalName, {
            repositoryName: this.repoName,
        })
    }

    createOutputs = () => {
        new Output(this.stack, 'RepositoryHttpUrl', {
            value: this.repository.repositoryCloneUrlHttp,
            disableExport: true
        })
    }
}