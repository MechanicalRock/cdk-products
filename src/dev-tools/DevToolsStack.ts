import { Stack, App } from '@aws-cdk/cdk';
import { CodeRepository } from './CodeRepository';
import { CodePipeline } from './CodePipeline';

export class DevToolsStack extends Stack {

    constructor(app: App, stackName: string) {
        super(app, stackName);
        this.create();
    }

    create = () => {
        // Create a code commit repo
        const repository = new CodeRepository(this, 'SampleStaticSiteRepo', 'sample-static-site').repository;

        // Create a code pipeline
        new CodePipeline({
            stack: this,
            logicalName: 'CodePipeline',
            pipelineName: 'code-pipeline',
            codeRepository: repository
        });
    }
}