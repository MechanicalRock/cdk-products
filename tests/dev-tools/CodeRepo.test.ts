import { Stack } from '@aws-cdk/cdk';
import * as yaml from '../utils/Yaml';
import { CodeRepository } from '../../src/dev-tools/CodeRepository';

describe('CodePipeline', () => {
    it('should match the template snapshot', () => {
        // Given
        const stack = new Stack();

        // When
        new CodeRepository(stack, 'CodeRepo', 'code-repo');

        // Then
        expect(yaml.toYamlString(stack.toCloudFormation())).toMatchSnapshot();
    })
})