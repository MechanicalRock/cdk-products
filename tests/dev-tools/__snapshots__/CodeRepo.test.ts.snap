// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodePipeline should match the template snapshot 1`] = `
"Resources:
  CodeRepoDC8A9D5F:
    Type: 'AWS::CodeCommit::Repository'
    Properties:
      RepositoryName: code-repo
      Triggers: []
Outputs:
  RepositoryHttpUrl:
    Value:
      'Fn::GetAtt':
        - CodeRepoDC8A9D5F
        - CloneUrlHttp
"
`;
